db/document.go

```diff
@@ -881,6 +886,49 @@ func (doc *Document) updateChannelHistory(channelName string, seq uint64, additi
 
 }
 
+func (doc *Document) addToChannelSetHistory(channelName string, historyEntry ChannelSetEntry) {
+	// Before adding the entry we need to verify the number of items in the document channel history so verify
+	// whether we need to prune them.
+	// As we iterate over the existing channels we will keep track of the oldest and second oldest entries along with
+	// their location in the slice. If we need to then prune we can 'merge' the oldest and second oldest and remove
+	// the oldest.
+
+	var oldestEntryStartSeq uint64 = math.MaxUint64
+	var secondOldestEntryStartSeq uint64 = math.MaxUint64
+
+	oldestEntryIdx := -1
+	secondOldestEntryIdx := -1
+
+	entryCount := 0
+
+	for entryIdx, entry := range doc.ChannelSetHistory {
+		if entry.Name == channelName {
+			entryCount++
+
+			if entry.Start < oldestEntryStartSeq {
+				secondOldestEntryStartSeq = oldestEntryStartSeq
+				oldestEntryStartSeq = entry.Start
+
+				secondOldestEntryIdx = oldestEntryIdx
+				oldestEntryIdx = entryIdx
+				continue
+			}
+
+			if entry.Start < secondOldestEntryStartSeq {
+				secondOldestEntryStartSeq = entry.Start
+				secondOldestEntryIdx = entryIdx
+			}
+		}
+	}
+
+	if entryCount == DocumentHistoryMaxEntriesPerChannel {
+		doc.ChannelSetHistory[secondOldestEntryIdx].Start = oldestEntryStartSeq
+		doc.ChannelSetHistory = append(doc.ChannelSetHistory[:oldestEntryIdx], doc.ChannelSetHistory[oldestEntryIdx+1:]...)
```

Comment:
Isn't this going to prune away entries for other channels if they were between `oldestEntryIdx` and `secondOldestEntryIdx`?

e.g:
```golang
doc.ChannelSetHistory = {
    {"a", 1, 5},
    {"b", 5},
    {"a", 7, 15},
    {"a", 20},
}
```

will merge seq ranges a:1-5 and a:7-15 as a:1-15 but then accidentally delete the entry for b

Reply:
Should we be using the memory safe version of a slice cut here?  https://github.com/golang/go/wiki/SliceTricks

Reply:
Don't think so. It doesn't contain any pointers so the current method should suffice with no memory leak issues.